### PR TITLE
Add required multiplayer models (and associated flows) for match "types"

### PR DESCRIPTION
**Pull request by peppy**
This adds new models which are to be used for client-server and server-client communication of events and states for modular match "type" implementations.

This PR intentionally doesn't add the ability to change to team VS, as it is intended that this can be merged in its current state without regressing any existing behaviours. PRs with the UI pieces to allow changing the mode (and teams) will come later today.

~~We should probably decide on what will happen to the naming before getting too far in review / dependency chains..~~

- [x] Depends on https://github.com/ppy/osu/pull/14112

### CODE REVIEW

**Comment by bdach** [comment]
First and foremost, re: naming - as already stated on discord I agree that "match ruleset" will be a potentially misleading name, but the worse thing is that the alternative terms I can think of are mostly generlc which doesn't help...

Is there an overarching vision of what future "match rulesets" there may be? So far we have:

1. ruleset (osu!, catch, taiko, mania) - what the basic rules of gameplay are
2. "match ruleset" (head to head, team vs., maybe coop later?) - _seems_ to mostly be about how players are configured/split among each other in the room
3. category ("normal"/playlist, spotlight, realtime) - how the resulting score is collected / what is its significance

If the description of (2) sounds like it will fit things going forward, then I came up with a few alternatives, although they're all still mostly generic ("match/room format"? "player/team configuration"? "multiplayer mode"?)

File: osu.Game/Online/API/IAPIProvider.cs [hunk 1]
```diff
@@ -102,7 +102,8 @@ public interface IAPIProvider
         /// </summary>
         /// <param name="clientName">The name of the client this connector connects for, used for logging.</param>
         /// <param name="endpoint">The endpoint to the hub.</param>
-        IHubClientConnector? GetHubConnector(string clientName, string endpoint);
+        /// <param name="preferMessagePack"></param>
```
Comment (bdach) [inline osu.Game/Online/API/IAPIProvider.cs:105]:
```suggestion
        /// <param name="preferMessagePack">Whether to use MessagePack for serialisation if available on this platform.</param>
```

File: osu.Game/Online/HubClientConnector.cs [hunk 2]
```diff
@@ -144,13 +147,22 @@ private HubConnection buildConnection(CancellationToken cancellationToken)
...
                 builder.AddMessagePackProtocol();
             else
             {
                 // eventually we will precompile resolvers for messagepack, but this isn't working currently
                 // see https://github.com/neuecc/MessagePack-CSharp/issues/780#issuecomment-768794308.
-                builder.AddNewtonsoftJsonProtocol(options => { options.PayloadSerializerSettings.ReferenceLoopHandling = ReferenceLoopHandling.Ignore; });
+                builder.AddNewtonsoftJsonProtocol(options =>
+                {
+                    options.PayloadSerializerSettings.ReferenceLoopHandling = ReferenceLoopHandling.Ignore;
+                    options.PayloadSerializerSettings = new JsonSerializerSettings
```
Comment (bdach) [inline osu.Game/Online/HubClientConnector.cs:159]: Looks as if the value set for `ReferenceLoopHandling` will be immediately discarded this way...?

File: osu.Game/Online/Multiplayer/MultiplayerClient.cs [hunk 3]
```diff
@@ -420,6 +422,46 @@ Task IMultiplayerClient.UserStateChanged(int userId, MultiplayerUserState state)
...
+                return Task.CompletedTask;
+
+            Scheduler.Add(() =>
+            {
+                if (Room == null)
+                    return;
+
+                Room.MatchRulesetState = state;
+                RoomUpdated?.Invoke();
+            });
```
Comment (bdach) [inline osu.Game/Online/Multiplayer/MultiplayerClient.cs:454]: This doesn't specify `forceScheduled = false` like `MatchRulesetUserStateChanged` does above - is that intentional or an oversight?

File: osu.Game/Online/Multiplayer/MultiplayerRoomSettings.cs [hunk 4]
```diff
@@ -39,6 +39,9 @@ public class MultiplayerRoomSettings : IEquatable<MultiplayerRoomSettings>
         [Key(7)]
         public string Password { get; set; } = string.Empty;
 
+        [Key(8)]
+        public MatchRulesetType MatchRulesetType { get; set; }
```
Comment (bdach) [inline osu.Game/Online/Multiplayer/MultiplayerRoomSettings.cs:43]: As a passing thought: this may need to become something more at some point if further configurability is to be desired (i.e. more teams, customisable team names, etc.) - as I understand it right now the server is in its entirety authoritative as to how the "multi ruleset" is enforced. I presume that is out of scope for now at least (if not at all).

**Review by bdach — DISMISSED** [review]

**Comment by smoogipoo** [comment]
@bdach There's also another concept to think of a name for, which is like... "playlist mode"? Options: `manual, karaoke/enqueued/"host-rotate", tournament`.

File: osu.Game/Online/Multiplayer/MultiplayerRoomSettings.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Online/Multiplayer/MultiplayerRoomSettings.cs:43]: Configuration is intended to be done via the event flow, not this enum.

**Comment by peppy** [comment]
> @bdach There's also another concept to think of a name for, which is like... "playlist mode"? Options: `manual, karaoke/enqueued/"host-rotate", tournament`.

I'd probably call this `ProgressionMode` or `ContinuationMode`.

File: osu.Game/Online/Multiplayer/MultiplayerClient.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Online/Multiplayer/MultiplayerClient.cs:454]: Seems like an oversight. For now I'm just trying to imitate the existing way things are done.

File: osu.Game/Online/Multiplayer/MultiplayerRoom.cs [hunk 5]
```diff
@@ -39,14 +39,17 @@ public class MultiplayerRoom
         /// All users currently in this room.
         /// </summary>
         [Key(3)]
-        public List<MultiplayerRoomUser> Users { get; set; } = new List<MultiplayerRoomUser>();
+        public IList<MultiplayerRoomUser> Users { get; set; } = new List<MultiplayerRoomUser>();
```
Comment (smoogipoo) [inline osu.Game/Online/Multiplayer/MultiplayerRoom.cs:42]: Is there a reason for this change?

File: osu.Game/Online/Multiplayer/MultiplayerRoom.cs [hunk 6]
```diff
@@ -39,14 +39,17 @@ public class MultiplayerRoom
...
 
         /// <summary>
         /// The host of this room, in control of changing room settings.
         /// </summary>
         [Key(4)]
         public MultiplayerRoomUser? Host { get; set; }
 
+        [Key(5)]
+        public MatchRoomState? MatchState { get; set; }
```
Comment (smoogipoo) [inline osu.Game/Online/Multiplayer/MultiplayerRoom.cs:51]: Not sold on the nullability (here and the user's state), but I guess we can change this at any point.

File: osu.Game/Online/Multiplayer/OnlineMultiplayerClient.cs [hunk 7]
```diff
@@ -56,6 +58,8 @@ private void load(IAPIProvider api)
                     connection.On(nameof(IMultiplayerClient.ResultsReady), ((IMultiplayerClient)this).ResultsReady);
                     connection.On<int, IEnumerable<APIMod>>(nameof(IMultiplayerClient.UserModsChanged), ((IMultiplayerClient)this).UserModsChanged);
                     connection.On<int, BeatmapAvailability>(nameof(IMultiplayerClient.UserBeatmapAvailabilityChanged), ((IMultiplayerClient)this).UserBeatmapAvailabilityChanged);
+                    connection.On<MatchRoomState>(nameof(IMultiplayerClient.MatchRoomStateChanged), ((IMultiplayerClient)this).MatchRoomStateChanged);
+                    connection.On<int, MatchUserState>(nameof(IMultiplayerClient.MatchUserStateChanged), ((IMultiplayerClient)this).MatchUserStateChanged);
```
Comment (smoogipoo) [inline osu.Game/Online/Multiplayer/OnlineMultiplayerClient.cs:62]: Any reason for not binding `MatchEvent` also?

File: osu.Game/Online/Multiplayer/MultiplayerRoom.cs (hunk 5 above)
Comment (peppy) [reply osu.Game/Online/Multiplayer/MultiplayerRoom.cs:42]: This is used server-side (see `ServerMultiplayerRoom`).

File: osu.Game/Online/Multiplayer/OnlineMultiplayerClient.cs (hunk 7 above)
Comment (peppy) [reply osu.Game/Online/Multiplayer/OnlineMultiplayerClient.cs:62]: Oversight (also unused for now, which is why it didn't come up in testing). Will add it in.

File: osu.Game/Online/Multiplayer/MultiplayerRoom.cs (hunk 5 above)
Comment (smoogipoo) [reply osu.Game/Online/Multiplayer/MultiplayerRoom.cs:42]: Yikes! Just to handle user joins/parts? That seems very overengineered.

Comment (peppy) [reply osu.Game/Online/Multiplayer/MultiplayerRoom.cs:42]: It was the simplest way I could think of, but I'm definitely open to alternatives (I brought this up in conversation on discord as I wasn't sure about it).

Comment (smoogipoo) [reply osu.Game/Online/Multiplayer/MultiplayerRoom.cs:42]: I guess this isn't a breaking change so can be reverted at any point, but I'll comment on the server-side usages.

**Review by smoogipoo — APPROVED** [review]
